File: benchmarks/json_parse.c
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>

static const char kData[] =
    "{\"users\":[{\"id\":1,\"name\":\"alice\",\"score\":12},"
    "{\"id\":2,\"name\":\"bob\",\"score\":7},"
    "{\"id\":3,\"name\":\"carol\",\"score\":42}],"
    "\"active\":true,\"count\":3,"
    "\"misc\":{\"neg\":-7,\"escape\":\"a\\n\\\"b\\\"\",\"unicode\":\"\\u0041\",\"zip\":90210},"
    "\"tags\":[\"a\",\"b\",\"c\"]}";

static inline bool isDigit(uint8_t ch) { return ch >= (uint8_t)'0' && ch <= (uint8_t)'9'; }

int main(void) {
  const int64_t outer = 20000;
  const int64_t len = (int64_t)(sizeof(kData) - 1);

  int64_t tokenTotal = 0;
  int64_t numberCount = 0;
  int64_t numberSum = 0;
  int64_t stringBytes = 0;

  for (int64_t o = 0; o < outer; ++o) {
    int64_t localTokens = 0;
    int64_t localNumberCount = 0;
    int64_t localNumberSum = 0;
    int64_t localStringBytes = 0;

    bool inString = false;
    bool escape = false;
    int32_t unicodeDigits = 0;
    bool inNumber = false;
    int64_t sign = 1;
    int64_t value = 0;

    for (int64_t i = 0; i < len; ++i) {
      uint8_t ch = (uint8_t)kData[i];

      if (inString) {
        if (unicodeDigits > 0) {
          unicodeDigits -= 1;
          continue;
        }
        if (escape) {
          escape = false;
          if (ch == (uint8_t)'u') {
            unicodeDigits = 4;
          }
          localStringBytes += 1;
          continue;
        }
        if (ch == (uint8_t)'\\') {
          escape = true;
          continue;
        }
        if (ch == (uint8_t)'"') {
          inString = false;
          continue;
        }
        localStringBytes += 1;
        continue;
      }

      if (inNumber) {
        if (isDigit(ch)) {
          value = value * 10 + (int64_t)(ch - (uint8_t)'0');
          continue;
        }
        localNumberSum += sign * value;
        localNumberCount += 1;
        inNumber = false;
        sign = 1;
        value = 0;
      }

      if (ch == (uint8_t)'"') {
        inString = true;
      } else if (ch == (uint8_t)'-') {
        inNumber = true;
        sign = -1;
        value = 0;
      } else if (isDigit(ch)) {
        inNumber = true;
        sign = 1;
        value = (int64_t)(ch - (uint8_t)'0');
      } else if (ch == (uint8_t)'{' || ch == (uint8_t)'}' || ch == (uint8_t)'[' || ch == (uint8_t)']' ||
                 ch == (uint8_t)':' || ch == (uint8_t)',') {
        localTokens += 1;
      }
    }

    if (inNumber) {
      localNumberSum += sign * value;
      localNumberCount += 1;
    }

    tokenTotal += localTokens;
    numberCount += localNumberCount;
    numberSum += localNumberSum;
    stringBytes += localStringBytes;
  }

  const int64_t output = numberSum + (tokenTotal * 1000000000LL) + (numberCount * 1000000LL) + stringBytes;
  printf("%lld\n", (long long)output);
  return 0;
}

File: benchmarks/json_parse.rs
const DATA: &str = r#"{"users":[{"id":1,"name":"alice","score":12},{"id":2,"name":"bob","score":7},{"id":3,"name":"carol","score":42}],"active":true,"count":3,"misc":{"neg":-7,"escape":"a\n\"b\"","unicode":"\u0041","zip":90210},"tags":["a","b","c"]}"#;

#[inline]
fn is_digit(ch: u8) -> bool {
    ch >= b'0' && ch <= b'9'
}

fn main() {
    let outer: i64 = 20_000;
    let bytes = DATA.as_bytes();

    let mut token_total: i64 = 0;
    let mut number_count: i64 = 0;
    let mut number_sum: i64 = 0;
    let mut string_bytes: i64 = 0;

    for _ in 0..outer {
        let mut local_tokens: i64 = 0;
        let mut local_number_count: i64 = 0;
        let mut local_number_sum: i64 = 0;
        let mut local_string_bytes: i64 = 0;

        let mut in_string = false;
        let mut escape = false;
        let mut unicode_digits: i32 = 0;
        let mut in_number = false;
        let mut sign: i64 = 1;
        let mut value: i64 = 0;

        for &ch in bytes {
            if in_string {
                if unicode_digits > 0 {
                    unicode_digits -= 1;
                    continue;
                }
                if escape {
                    escape = false;
                    if ch == b'u' {
                        unicode_digits = 4;
                    }
                    local_string_bytes += 1;
                    continue;
                }
                if ch == b'\\' {
                    escape = true;
                    continue;
                }
                if ch == b'"' {
                    in_string = false;
                    continue;
                }
                local_string_bytes += 1;
                continue;
            }

            if in_number {
                if is_digit(ch) {
                    value = value * 10 + (ch - b'0') as i64;
                    continue;
                }
                local_number_sum += sign * value;
                local_number_count += 1;
                in_number = false;
                sign = 1;
                value = 0;
            }

            if ch == b'"' {
                in_string = true;
            } else if ch == b'-' {
                in_number = true;
                sign = -1;
                value = 0;
            } else if is_digit(ch) {
                in_number = true;
                sign = 1;
                value = (ch - b'0') as i64;
            } else if ch == b'{' || ch == b'}' || ch == b'[' || ch == b']' || ch == b':' || ch == b',' {
                local_tokens += 1;
            }
        }

        if in_number {
            local_number_sum += sign * value;
            local_number_count += 1;
        }

        token_total += local_tokens;
        number_count += local_number_count;
        number_sum += local_number_sum;
        string_bytes += local_string_bytes;
    }

    let output: i64 = number_sum + (token_total * 1_000_000_000) + (number_count * 1_000_000) + string_bytes;
    println!("{}", output);
}

